src/services/services.ts

```diff
@@ -6772,6 +6762,67 @@ namespace ts {
             return [];
         }
 
+        /**
+         * Checks if position points to a valid position to add JSDoc comments, and if so,
+         * returns the appropriate scaffolding. Otherwise returns an empty string.
+         * Valid positions are
+         * * outside of comments, statements, and expressions, and
+         * * preceding a function declaration.
+         * 
+         * In VS, we additionally check that:
+         * * The line is all whitespace up to 'position' before performing the insertion.
+         * * If the keystroke sequence "/\*\*" induced the call, we also check that the next
+         * non-whitespace character is '*', which (approximately) indicates whether we added 
+         * the second '*' to complete an existing (JSDoc) comment.
+         * @param fileName The file in which to perform the check.
+         * @param position The (character-indexed) position in the file where the check should
+         * be performed.
+         */
+        function getDocCommentScaffoldingAtPosition(fileName: string, position: number): TextInsertion {
+            // Indicates the position is not appropriate to insert a comment (eg: a string or a regex).
+            const nullResult: TextInsertion = { newText: "", cursorOffset: 0 };
+            let start = new Date().getTime();
+            let sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+            log("getDocCommentScaffoldingAtPosition: getCurrentSourceFile: " + (new Date().getTime() - start));
+
+            // Check if in a context where we don't want to perform any insertion
+            if (isInString(sourceFile, position) || isInComment(sourceFile, position) || hasDocComment(sourceFile, position)) {
+                return nullResult;
+            }
+
+            let nodeAtPos = getTokenAtPosition(sourceFile, position);
+            let nodeStart = nodeAtPos.getStart()
+            if (!nodeAtPos || nodeStart < position) {
+                return nullResult;
+            }
+
+            let containingFunction = <FunctionDeclaration>getAncestor(nodeAtPos, SyntaxKind.FunctionDeclaration);
+
+            if (!containingFunction || containingFunction.getStart() < position) {
+                return nullResult;
+            }
+
+            let parameters = containingFunction.parameters;
+            let posLineAndChar = sourceFile.getLineAndCharacterOfPosition(position);
+            let lineStart = sourceFile.getLineStarts()[posLineAndChar.line];
+
+            let indentationStr = sourceFile.text.substr(lineStart, posLineAndChar.character).match(/\s*/).toString();
+
+            let docParams = parameters.map((p, index) =>
+                indentationStr + " * @param " + (p.name.kind === SyntaxKind.Identifier ? (<Identifier>p.name).text : "param" + index.toString()) + "\n");
+
+            let result = 
+                /* opening comment */               "/**\n" + 

```

Comment:
Changed -- consolidated the comments above the code. Though I disagree about the readability, a formatter would wreck it.